cmd/lp-colouring/colouring.go

```diff
@@ -0,0 +1,180 @@
+package main
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ScottSallinen/lollipop/graph"
+)
+
+const EmptyColour = math.MaxUint32
+
+type VertexProperty struct {
+	NbrColours sync.Map
+	WaitCount  int64
+	Colour     uint32
+}
+
+func (p *VertexProperty) String() string {
+	s := fmt.Sprintf("{%d,%d,[", p.Colour, p.WaitCount)
+	p.NbrColours.Range(func(key, value any) bool {
+		s += fmt.Sprintf("%d:%d,", key, value)
+		return true
+	})
+	return s + "]}"
+}
+
+func hash(id uint32) (hash uint32) {
+	// TODO: dummy hash function
+	return id
+}
+
+func comparePriority(p1, p2 uint32, id1, id2 uint32) bool {
+	return p1 > p2 || (p1 == p2 && id1 > id2)
+}
+
+// findFirstUnused finds the smallest unused colour, modifies colours, O(n)
+func findFirstUnused(colours []uint32) (firstUnused uint32) {
```

Comment:
Nice attempt! This definitely looks like it will work, but it is generic and thus not the most efficient. It may be O(n), but we can do better (use less memory and better constants for the O(N)).

Because we have constraints we can take advantage of, we can do this more efficiently.
1: There must be a free colour in range of our vertex degree.
2: We do not care if a neighbour has a colour > our degree. (nor do we care if multiple neighbours have the same colour)

Because of 1, we only need a bitmap representing the colour space, of degree size (len(outedges))
Because of 2, when looking through neighbour colours, we simply just need to set the index of the used colour in the bitmap to 1. If their colour is outside our degree, we don't need to do anything as we have no interest in it.

Then, we can look for the first zero bit and immediately return that index as the first unused colour.
Note: offset index -> colour by 1 as we don't care about colour 0, i.e., index 0 means colour 1.

(Or, better, do the opposite, initialize the bitmap as 1's and set 0's as used indices, making a flipped bitmap. Then the first index of a set bit is the unused colour, and this problem is known as "Find first set" and has hardware intrinsics. Doesn't seem Go's builtin bitmap has this, but I guess you could do it with something like https://pkg.go.dev/math/bits#TrailingZeros64 or find someone else's bitmap implementation, like https://github.com/kelindar/bitmap maybe)

Reply:
https://github.com/ScottSallinen/lollipop/pull/1/commits/2f2be8008738bf27624116e0969971b8a17ec122

Looks like the bitmap implementation there has what we need!
We'll let the bitmap be dynamic to resolve the issue we discussed about potential size mismatch.

Reply:
Interesting! If we allow the bitmap to grow, we are unable to utilize 2. Which way is more efficient for large graphs?

Reply:
This will still be more efficient, we only might use extra space; yet the average case for maximum colour is low, and a bitmap will be no larger than 1/8th the max colour ever used, in bytes.
Ideally we will discard the entries above our degree, but there is that odd case where we may have more entries in our map than in our degree (an artifact of skipping the queue, as we discussed).
I may have an idea for fixing that issue, the we can fix this.
Might be worth leaving a comment about discarding an entry > degree in the future.

Reply:
Got it! Thanks.